hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -1843,20 +1849,151 @@ func (r *HostedClusterReconciler) reconcile(ctx context.Context, req ctrl.Reques
+		// against the one from the secret
+		hashFromServerEndpoint, err := kasServingCertHashFromEndpoint()
+		if err != nil {
+			return nil, err
+		}
+		// If they match, we still need to store the hash for future comparisons
+		if hashFromServerEndpoint == hashFromSecret {
+			return map[string]string{
+				kasServingCertHashAnnotation: hashFromSecret,
+			}, nil
+		}
+		// If they don't match, we store the hash AND initiate a restart
+		log.Info("WARNING: The KAS endpoint server certificate does not match the serving certificate secret. Setting the annotation to restart control plane workloads.")
+		return map[string]string{
+			kasServingCertHashAnnotation:  hashFromSecret,
+			hyperv1.RestartDateAnnotation: fmt.Sprintf("CertHash:%s", hashFromSecret),
+		}, nil
+	}
+}
+
+func kasHostAndPortFromHCP(hcp *hyperv1.HostedControlPlane) string {
+	// NOTE: On IBM Cloud platform, the service port is different than the default. However, in that platform
+	// PKI is not managed by the CPO, therefore this check would never run. For simplicity, we only use the default
+	// port here.
+	return fmt.Sprintf("kube-apiserver.%s.svc:%d", hcp.Namespace, config.KASSVCPort)
+}
+
+func (r *HostedClusterReconciler) kasServingCertHashFromSecret(ctx context.Context, hcp *hyperv1.HostedControlPlane) func() (string, error) {
+	return func() (string, error) {
+		servingCertSecret := manifests.KASServingCertSecret(hcp.Namespace)
+		if err := r.Get(ctx, client.ObjectKeyFromObject(servingCertSecret), servingCertSecret); err != nil {
+			return "", err
+		}
+		if value := servingCertSecret.Data[corev1.TLSCertKey]; len(value) == 0 {
+			return "", fmt.Errorf("no value for KAS serving certificate in %s", client.ObjectKeyFromObject(servingCertSecret).String())
+		} else {
+			return hyperutil.HashSimple(value), nil
+		}
+	}
+}
+
+func (r *HostedClusterReconciler) kasServingCertHashFromEndpoint(kasHostAndPort string) func() (string, error) {
+	return func() (string, error) {
+		conn, err := tls.Dial("tcp", kasHostAndPort, &tls.Config{
+			InsecureSkipVerify: true,
+			ServerName:         "kubernetes",
+		})
+		if err != nil {
+			return "", fmt.Errorf("failed to dial %s: %w", kasHostAndPort, err)
+		}
+		defer conn.Close()
+		kasCerts := conn.ConnectionState().PeerCertificates
+		if len(kasCerts) == 0 {
+			return "", fmt.Errorf("no certificate found on KAS endpoint %s", kasHostAndPort)
+		}
+		pemBytes := certs.CertToPem(kasCerts[0])
+		return hyperutil.HashSimple(pemBytes), nil
+	}
+}
+
+// shouldCheckForStaleCerts returns true if a HostedCluster should be checked for stale certs
+// The following pre-conditions must be met:
+// - The HostedCluster must be managing its own PKI (the hyperv1.DisablePKIReconciliationAnnotation is not present)
+// - The HostedCluster has been available once
+// - The cluster version of the HostedCluster is < 4.19.0
+var version419 = semver.MustParse("4.19.0")
```

Comment:
There's an exportable var of this already in version.go. I wonder if we should add this to config.go and update both 🤔

Reply:
The one in version.go should change as we move the main branch to a new version, so I can't rely on that. I will add it to config.go